### PR TITLE
feat: implement recursive placeholder resolution in Default parameters

### LINKED ISSUE #1009: Allow ${variables} in Default parameters
It'd be great to be able to init <Default> parameters with ${variables}, like ${libdir} or ${ipmdir}. It can be very helpful while installing packages on different IRIS servers, and let packages behave according to these parameters.
I tried the following, and it didn't work :
```xml
<Default Name="datapath" Value="${libdir}data/" />
```
it ended up with the following error:
```
<REGULAR EXPRESSION>ReplaceAll+3^%Regex.Matcher.1
```

### PR DESCRIPTION
**Pull request by AshokThangavel**
### **Description**
This PR introduces deeply nested variables or chained configuration settings where one `${variable}` depends on another.
 with a safety handle complex dependencies and prevent infinite loops in the case of circular references.

fixes: #1009

### **Changes**

* **Recursive Logic:** Updated `ResolvePlaceholders` to use a `while` loop that continues as long as substitutions are being made.
* **Efficiency:** The loop exits immediately if a pass results in no changes, ensuring no performance penalty for simple configurations.

### **Testing Performed**

1. **Unit Tests:** Verified with complex nested strings:
* `${A}` -> `${B}` -> `${C}` -> `FinalValue`.

2. **Integration Test:** Added a new integration test class `Test.PM.Integration.Module` that:
* Generates a `module.xml` from XData.
* Writes it to the filesystem.
* Executes an IPM `load` command to verify the full lifecycle of the variable resolution.
### Test Execution Results: `Test.PM.Integration.Module`
| Counter | Action | Status | Description |
| --- | --- | --- | --- |
| 1 | LogMessage | passed | start Loading the demo-module1 module |
| 2 | AssertStatusOK | passed | Directory created /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 3 | AssertStatusOK | passed | module.xml File created on /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 4 | AssertStatusOK | passed | Created the xml file on /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 5 | AssertStatusOK | passed | Loaded varresolver module successfully from /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 6 | AssertTrue | passed | Module demo-module1 exists in IPM and version is 1.0.0 |
| 7 | LogMessage | passed | List all modules |
| 8 | AssertStatusOK | passed | uninstalled module demo-module1 successfully. |
| 9 | AssertStatusOK | passed | Deleted the module.xml file from /home/irisowner/zpm/tests/integration_tests/Test/PM/Integration/_data/varresolver/ |
| 10 | LogMessage | passed | Duration of execution: 1.193466 sec. |

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - #822: The CPF resource processor now supports system expressions and macros in CPF merge files
 - #578 Added functionality to record and display IPM history of install, uninstall, load, and update
 - #961: Adding creation of a lock file for a module by using the `-create-lockfile` flag on install.
+- #1013: Implement recursive placeholder resolution in Default parameters
```
Comment (isc-dchui) [inline CHANGELOG.md]: Move this to the new 0.10.6 section please

**Review by isc-dchui — CHANGES_REQUESTED** [review]
A few comments and questions

File: tests/integration_tests/Test/PM/Integration/Module.cls [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			<Default Name="datadefaultmgrdir" Value="${mgrdir}/" />
+			<Default Name="mTestPlaceHolder" Value="${dataDefaultTest2}/xtest/${version}/${mgrdir}/${mTestVersion}/${datapath1}" />
+			<Default Name="datadefaultcspdir" Value="${cspdir}/" />
+			<Default Name="dataDefaultTest2" Value="${datadefaultcspdir}/xdata" />
+			<Default Name="dataDefaultTest3" Value="${dataDefaultTest2}/xtest" />
+			<Default Name="mTestVersion" Value="${dataDefaultTest2}/xtest/${version}" />
+			<Default Name="ipmtest" Value="TESTING MY STRING"/>
+			<Default Name="ipmdir" Value="/usr/irissys/mgr/user/mts"/>
+			<Default Name="datapath" Value="${libdir}data/" />
+			<Default Name="datapath1" Value="${ipmdir}data/" />
```
[the remark this answers is not in the document]
Comment (AshokThangavel) [reply tests/integration_tests/Test/PM/Integration/Module.cls]: Thank you for pointing out that excellent scenario. You are absolutely right—this is a 'pathological' but critical edge case for configuration flexibility. I have updated the implementation.

File: src/cls/IPM/Storage/ModuleSetting/Default.cls [hunk 3]
```diff
@@ -57,6 +57,40 @@ ClassMethod EvaluateAttribute(
...
+        set maxLevels = maxLevels - 1
+        set param = ""
+
+        for {
+            set param = $order(customParams(param), 1, data)
+            quit:param=""
+            continue:data'["${"
+
+            set varExpr = "${" _ $piece($piece(data, "${", 2), "}") _ "}"
+            set resolved = ##class(%IPM.Utils.Module).%EvaluateSystemExpression(varExpr)
```
[the remark this answers is not in the document]
Comment (AshokThangavel) [reply src/cls/IPM/Storage/ModuleSetting/Default.cls]: I've modified the logic to ensure the system variables and custom variables.

File: tests/integration_tests/Test/PM/Integration/Module.cls [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    do $$$AssertStatusOK(status,"Created the xml file on "_moduleDir)
+
+    set status = ##class(%IPM.Main).Shell("load "_moduleDir)
+    do $$$AssertStatusOK(status,"Loaded "_..#CommonPathPrefix_" module successfully from "_moduleDir)
+
+    set module = ##class(%IPM.Storage.Module).NameOpen(..#ModuleName)
+    do $$$AssertTrue($isobject(module), "Module "_..#ModuleName_" exists in IPM and version is "_ module.Version.ToString())
+
+        do $$$LogMessage("List all modules")
+    set status = ##class(%IPM.Main).Shell("list")
```
[the remark this answers is not in the document]
Comment (AshokThangavel) [reply tests/integration_tests/Test/PM/Integration/Module.cls]: Thank you!. I've added the `$$$AssertStatusOK`

File: src/cls/IPM/Storage/ModuleSetting/Default.cls [hunk 5]
```diff
@@ -57,6 +57,40 @@ ClassMethod EvaluateAttribute(
     return attribute
 }
 
+ClassMethod ResolvePlaceholders(ByRef customParams)
```
[the remark this answers is not in the document]
Comment (AshokThangavel) [reply src/cls/IPM/Storage/ModuleSetting/Default.cls]: Thank you. Added the comments on top of method `ResolvePlaceholders`

**Review by isc-dchui — COMMENTED** [review]
Thanks for the updates! Here's another pass of comments/questions

**Review by isc-dchui — COMMENTED** [review]
Just a few more things, thanks!

**Comment by AshokThangavel** [comment]
I’ve pulled the latest changes from main and resolved the conflicts in CHANGELOG.md. The new entry has been updated to `0.10.7`. Thank you!

**Review by isc-dchui — DISMISSED** [review]

**Review by isc-dchui — DISMISSED** [review]

**Review by isc-tleavitt — APPROVED** [review]
